EIPS/eip-4931.md

```diff
@@ -0,0 +1,353 @@
+``` solidity
+/// @dev A method to mock the upgrade call determining the amount of destination tokens received from an upgrade
+/// as well as the amount of source tokens that are left over as remainder
+/// @param sourceAmount The amount of source tokens that will be upgraded
+/// @return destinationAmount A uint256 representing the amount of destination tokens received if upgrade is called
+/// @return sourceRemainder A uint256 representing the amount of source tokens left over as remainder if upgrade is called
+function computeUpgrade(uint256 sourceAmount) external view
+        returns (uint256 destinationAmount, uint256 sourceRemainder)
+```
+
+#### computeDowngrade (Optional Ext.)
+
+Computes the `sourceAmount` of source tokens that correspond to a given `destinationAmount` of destination tokens, according to the predefined conversion ratio, as well as the `destinationRemainder` amount of destination tokens that can't be downgraded. For example, let's consider a (3, 2) ratio, which means that 3 destination tokens are provided for every 2 source tokens; for a destination amount of 13 tokens, `computeDowngrade(13)` should return `(4, 1)`, meaning that 4 source tokens are expected (in this case, from 12 destination tokens) and 1 destination token is left as remainder.
+``` solidity
+/// @dev A method to mock the downgrade call determining the amount of source tokens received from a downgrade
+/// as well as the amount of destination tokens that are left over as remainder
+/// @param destinationAmount The amount of destination tokens that will be downgraded
+/// @return sourceAmount A uint256 representing the amount of source tokens received if downgrade is called
+/// @return destinationRemainder A uint256 representing the amount of destination tokens left over as remainder if upgrade is called
+function computeDowngrade(uint256 destinationAmount) external view
+        returns (uint256 sourceAmount, uint256 destinationRemainder)
+```
+
+
+#### upgrade
+
+Upgrades the `amount` of source token to the destination token in the specified ratio. The destination tokens will be sent to the ```_to``` address. The function MUST lock the source tokens in the upgrade contract or burn them. If the downgrade Optional Ext. is implemented, the source tokens MUST be locked instead of burning. The function MUST `throw` if the caller's address does not have enough source token to upgrade or if ```isUpgradeActive``` is returning ```false```. The function MUST also fire the `Upgrade` event. `approve` MUST be called first on the source contract.
```

Comment:
```suggestion
Upgrades the `amount` of source token to the destination token in the specified ratio. The destination tokens will be sent to the `_to` address. The function MUST lock the source tokens in the upgrade contract or burn them. If the downgrade Optional Ext. is implemented, the source tokens MUST be locked instead of burning. The function MUST `throw` if the caller's address does not have enough source token to upgrade or if `isUpgradeActive` is returning `false`. The function MUST also fire the `Upgrade` event. `approve` MUST be called first on the source contract.
```

You guessed it -- single backticks